File: PeekIterator.java
// Java Iterator interface reference:
// https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html

//Time -O(1)
//space - O(1)
class PeekingIterator implements Iterator<Integer> {
    Iterator<Integer> iterator;
    Queue<Integer> q;
    Integer prev;
    boolean hasPrev;
	public PeekingIterator(Iterator<Integer> iterator) {
	    // initialize any member here.
	    this.iterator= iterator;
        this.q = new LinkedList<>();
	}
	
    // Returns the next element in the iteration without advancing the iterator.
	public Integer peek() {
        if(iterator.hasNext() && hasPrev==false){
            prev = iterator.next();
            hasPrev = true;
        }
        return prev;
	}
	
	// hasNext() and next() should behave the same as in the Iterator interface.
	// Override them if needed.
	@Override
	public Integer next() {
	    if(hasPrev) {
            hasPrev = false;
            return prev;
        }
        
        return iterator.next();
	}
	
	@Override
	public boolean hasNext() {
	    if(hasPrev || iterator.hasNext()) return true;
        return false;
	}
}

File: PeekIterator.py
# Below is the interface for Iterator, which is already defined for you.
#
# class Iterator:
#     def __init__(self, nums):
#         """
#         Initializes an iterator object to the beginning of a list.
#         :type nums: List[int]
#         """
#
#     def hasNext(self):
#         """
#         Returns true if the iteration has more elements.
#         :rtype: bool
#         """
#
#     def next(self):
#         """
#         Returns the next element in the iteration.
#         :rtype: int
#         """

class PeekingIterator:
    # Time, Space Complexity: O(1), O(1)
    def __init__(self, iterator):
        """
        Initialize your data structure here.
        :type iterator: Iterator
        """
        self.nums = iterator
        #self.c = 0
        self.queue = None

    def peek(self):
        """
        Returns the next element in the iteration without advancing the iterator.
        :rtype: int
        """
        if(self.queue!=None):
            return self.queue
        if(self.nums.hasNext()):
            x = self.nums.next()
            self.queue = x
            return x

    def next(self):
        """
        :rtype: int
        """
        if(self.queue!=None):
            x = self.queue
            self.queue = None
            return x
        if(self.nums.hasNext()):
            x = self.nums.next()
            return x
        

    def hasNext(self):
        """
        :rtype: bool
        """
        if(self.queue!=None or self.nums.hasNext()):
            return True
        else:
            return False
        

# Your PeekingIterator object will be instantiated and called as such:
# iter = PeekingIterator(Iterator(nums))
# while iter.hasNext():
#     val = iter.peek()   # Get the next element but not advance the iterator.
#     iter.next()         # Should return the same value as [val].
